ginga/canvas/types/plots.py

```diff
@@ -0,0 +1,832 @@
+        separate arrays.
+        """
+        self.plot(np.asarray((xpts, ypts)).T)
+
+    def plot(self, points, limits=None):
+        """Plot `points`, a list, tuple or array of (x, y) points.
+
+        Parameter
+        ---------
+        points : array-like
+            list, tuple or array of (x, y) points
+
+        limits : array-like, optional
+            array of (xmin, ymin), (xmax, ymax)
+
+        Limits will be calculated if not passed in.
+        """
+        self.points = np.asarray(points)
+        self.plot_xlim = (None, None)
+
+        # set or calculate limits
+        if limits is not None:
+            # passing limits saves costly min/max calculation
+            self.limits = np.asarray(limits)
+        else:
+            self._calc_limits(self.points)
+
+    def _calc_limits(self, points):
+        """Internal routine to calculate the limits of `points`.
+        """
+        # TODO: what should limits be if there are no points?
+        if len(points) == 0:
+            self.limits = np.array([[0.0, 0.0], [0.0, 0.0]])
+        else:
+            x_vals, y_vals = points.T
+            self.limits = np.array([(x_vals.min(), y_vals.min()),
+                                    (x_vals.max(), y_vals.max())])
+
+    def calc_points(self, viewer, start_x, stop_x):
+        """Called when recalculating our path's points.
+        """
+        # in case X axis is flipped
+        start_x, stop_x = min(start_x, stop_x), max(start_x, stop_x)
+
+        new_xlim = (start_x, stop_x)
+        if new_xlim == self.plot_xlim:
+            # X limits are the same, no need to recalculate points
+            return
+
+        self.plot_xlim = new_xlim
+
+        points = self.get_data_points(points=self.points)
+        if len(points) == 0:
+            self.path.points = points
+            return
+
+        x_data, y_data = points.T
+
+        # if we can determine the visible region shown on the plot
+        # limit the points to those within the region
+        if np.all(np.isfinite([start_x, stop_x])):
+            idx = np.logical_and(x_data >= start_x, x_data <= stop_x)
+            points = points[idx]
+
+        if self.x_func is not None:
+            # now find all points position in canvas X coord
+            cpoints = self.get_cpoints(viewer, points=points)
+            cx, cy = cpoints.T
+
+            # Reduce each group of Y points that map to a unique X via a
+            # function that reduces to a single value.  The desirable function
+            # will depend on the function of the plot, but mean() would be a
+            # sensible default
+            _, i = np.unique(cx, return_index=True)
+            gr_pts = np.split(points, i)
+            x_data = np.array([self.x_func(a.T[0]) for a in gr_pts
+                               if len(a) > 0])
+            y_data = np.array([self.y_func(a.T[1]) for a in gr_pts
+                               if len(a) > 0])
```

Comment:
Is this too much code simply to avoid tranposing `a` twice?
```suggestion
            x_data = []
            y_data = []
            for a in gr_pts:
                if len(a) <= 0:
                    continue
                a_t = a.T
                x_data.append(self.x_func(a_t[0]))
                y_data.append(self.y_func(a_t[1]))

            x_data = np.array(x_data)
            y_data = np.array(y_data)
```

Reply:
I think the list comprehension is shorter and perhaps easier to understand. Probably faster than a Python loop with appends too.

Reply:
Depends on how expensive the transpose operation is. Hard to tell without having some realistic data to benchmark on my end. If you think this is premature optimization or even regression, feel free to ignore. Thanks!

Reply:
Pretty sure that numpy transpose (with .T) does not reorder data in memory, but only makes a new stride/offset for accessing the data. (https://stackoverflow.com/questions/19479384/is-numpy-transpose-reordering-data-in-memory).

Reply:
Okay then. I don't know why I remember transposing a cube was eating up all my processing time. Maybe the way I had to do it, it created a copy. 🤷